### PR TITLE
🐛 Fix XValidations flattening

### LINKED ISSUE #997: Invalid CRD generated when XValidation is used on both field and struct
When the `XValidation` marker is used on both a field and the type for that field the generated CRD is invalid because it tries to use an `allOf` property to merge the validations.

For example, given the definition of the following structs:
```
type FooSpec struct {
	// +kubebuilder:validation:XValidation:rule="size(self.field) > 2",message="validation 1"
	Thing Thing `json:"thing"`
}

// +kubebuilder:validation:XValidation:rule="has(self.field)",message="validation 2"
type Thing struct {
	Field *string `json:"field"`
}
```

The generated CRD will include:
```
properties:
  thing:
    allOf:
    - x-kubernetes-validations:
      - message: validation 2
        rule: has(self.field)
    - x-kubernetes-validations:
      - message: validation 1
        rule: size(self.field) > 2
    properties:
      field:
        type: string
    required:
    - field
    type: object
```

Where the expected result would have been:
```
properties:
  thing:
    x-kubernetes-validations:
    - message: validation 2
      rule: has(self.field)
    - message: validation 1
      rule: size(self.field) > 2
    properties:
      field:
        type: string
    required:
    - field
    type: object
```

I have a full reproduction of the bug including the invalid CRDs on https://github.com/cezarsa/validationbug/blob/a8148487551b92b4f3b6669cde9c4050e35cb98d/main.go.

### PR DESCRIPTION
**Pull request by cezarsa**
Fixes #997

This PR ensures XValidations are merged correctly when flattening the CRD schema.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Do we need to consider the ordering in flattening? CRD validations are executed in order so do we think the field level validations or the struct level validations should come first? What will happen with this approach, I assume the output here is deterministic?

**Comment by cezarsa** [comment]
> What will happen with this approach, I assume the output here is deterministic?

It is deterministic and, with the current approach, struct-level validations will come first. This kinda makes sense to me but I don't have a definitive answer if this is the intended behavior or if this has been discussed before.

Ultimately, I don't think the order can affect the correctness or the cost budget for the validations as they are all executed even after a failed rule.

**Comment by JoelSpeed** [comment]
> Ultimately, I don't think the order can affect the correctness or the cost budget for the validations as they are all executed even after a failed rule.

Is that true? I was under the impression that if a rule failed, further rules were not executed until the earlier rule passed?

Either way, I guess that doesn't hugely matter since all rules must pass for the object to be accepted. I would be tempted to poke some of the API machinery folks though just in case they can foresee any issues here/have a suggestion for prioritisation

**Comment by cezarsa** [comment]
> Is that true? I was under the impression that if a rule failed, further rules were not executed until the earlier rule passed?

Yes, I just tested it to be sure, using controller-gen from this branch and with an object declared as:
```
type FooSpec struct {
	// +kubebuilder:validation:XValidation:rule="false",message="validation 3"
	// +kubebuilder:validation:XValidation:rule="false",message="validation 4"
	Thing *Thing `json:"thing"`
}

// +kubebuilder:validation:XValidation:rule="false",message="validation 1"
// +kubebuilder:validation:XValidation:rule="false",message="validation 2"
type Thing struct {
	Field *string `json:"field"`
}
```

Applying it returned all the validation errors:
```
$  k apply -f ./example.yaml
customresourcedefinition.apiextensions.k8s.io/foos.validation.bug unchanged
The Foo "foo1" is invalid:
* spec.thing: Invalid value: "object": validation 1
* spec.thing: Invalid value: "object": validation 2
* spec.thing: Invalid value: "object": validation 3
* spec.thing: Invalid value: "object": validation 4
```

**Comment by sbueringer** [comment]
I think for cost estimation at CRD create/update all rules are compiled and afterwards error messages are produced accordingly: https://github.com/kubernetes/kubernetes/blob/d081fd56a2aba4663f11dd5e27dbcf64c0ec6638/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1226

During CR create/update:
* I"m not sure if having the allOff or not makes any difference: https://github.com/kubernetes/kubernetes/blob/b34197dc14aecdd1da1c70b979d1b304f522cc30/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L305-L318
  * Still agree though that not adding an unnecessary allOf is preferable
* It looks to me like rule evaluation will stop once the budget limit is hit: https://github.com/kubernetes/kubernetes/blob/b34197dc14aecdd1da1c70b979d1b304f522cc30/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L422
  * But I think as long as the order is the same, it doesn't matter if the rules are in one `x-validations` element or separate in multiple below an `allOf`

@cezarsa @JoelSpeed WDYT? does that make sense?

(feel free to cc someone from api-machinery, not sure who :))

File: pkg/crd/flatten_all_of_test.go [hunk 1]
```diff
@@ -241,6 +241,25 @@ var _ = Describe("AllOf Flattening", func() {
 				},
 			}))
 		})
+
+		It("should merge XValidation fields", func() {
```
Comment (sbueringer) [inline pkg/crd/flatten_all_of_test.go:245]: Thx for the unit test, let's please also add test coverage in https://github.com/kubernetes-sigs/controller-tools/blob/5e5bc88a01cab8ebff8c9fc4d1e90bb858139186/pkg/crd/testdata/cronjob_types.go

Comment (cezarsa) [reply pkg/crd/flatten_all_of_test.go:245]: Done! Added it on https://github.com/kubernetes-sigs/controller-tools/pull/998/commits/43611064a1be33cfab6d25aa1b03bd11a1b39d3c

**Comment by cezarsa** [comment]
> I'm not sure if having the allOf or not makes any difference:

Having `allOf` on validations in the CRD actually causes the apiserver to reject it. I tried applying the example CRD from https://github.com/kubernetes-sigs/controller-tools/issues/997 on a k8s v1.28.9 API and it failed with: 
```
$ kubectl apply -f ./crd/validation.bug_foos.yaml
The CustomResourceDefinition "foos.validation.bug" is invalid:
* spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[1].x-kubernetes-validations: Forbidden: must be empty to be structural
```

**Comment by jpbetz** [comment]
> > I'm not sure if having the allOf or not makes any difference:
> 
> Having `allOf` on validations in the CRD actually causes the apiserver to reject it. I tried applying the example CRD from #997 on a k8s v1.28.9 API and it failed with:
> 
> ```
> $ kubectl apply -f ./crd/validation.bug_foos.yaml
> The CustomResourceDefinition "foos.validation.bug" is invalid:
> * spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
> * spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[1].x-kubernetes-validations: Forbidden: must be empty to be structural
> ```

https://github.com/kubernetes/kubernetes/pull/124381 was merged into master quite recently. Does it resolve this problem?

**Comment by cezarsa** [comment]
> https://github.com/kubernetes/kubernetes/pull/124381 was merged into master quite recently. Does it resolve this problem?

Hey @jpbetz, thanks I wasn't aware of that PR but apparently it doesn't solve the problem.
I've just compiled apiserver from source and still see the same error. Repro steps:

1. Clone https://github.com/kubernetes/kubernetes
2.
        cd kubernetes
        make kube-apiserver kubectl
        ./hack/install-etcd.sh
        cp ./third_party/etcd/etcd _output/bin/
3. Clone https://github.com/cezarsa/validationbug
4.
        cd validationbug
        KUBEBUILDER_ASSETS=../kubernetes/_output/bin go test

The test above tries to apply a CRD with [`x-kubernetes-validations` inside a `all_of` block](https://github.com/cezarsa/validationbug/blob/b9fdc82b844526c238aaac733d8486230285e885/crd/validation.bug_foos.yaml#L41-L47) but it failed with:


```
--- FAIL: TestFunc (6.45s)
    main_test.go:16: unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "foos.validation.bug": CustomResourceDefinition.apiextensions.k8s.io "foos.validation.bug" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural, spec.validation.openAPIV3Schema.properties[spec].properties[thing].allOf[1].x-kubernetes-validations: Forbidden: must be empty to be structural]
FAIL
exit status 1
FAIL	github.com/cezarsa/validationbug	7.120s
```

File: pkg/crd/flatten.go [hunk 2]
```diff
@@ -147,6 +147,8 @@ func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps, e
 			dstField.Set(srcField)
 		case "XMapType":
 			dstField.Set(srcField)
+		case "XValidations":
+			dstField.Set(reflect.AppendSlice(dstField, srcField))
```
Comment (sbueringer) [inline pkg/crd/flatten.go:151]: I would flip the order here. I think it's better if the validation from the field is before the type.

Basically this allows to add validations add the start of the validation array when importing external types (~ that would prioritize "local" validations over imported ones)

**Comment by sbueringer** [comment]
@cezarsa Sorry for the delay. Last comment from my side: https://github.com/kubernetes-sigs/controller-tools/pull/998#discussion_r1710843042

I think even if the current output would be somehow valid (which it doesn't seem to be), it still makes sense to flatten in this case

File: pkg/crd/flatten.go (hunk 2 above)
Comment (cezarsa) [reply pkg/crd/flatten.go]: Order flipped in https://github.com/kubernetes-sigs/controller-tools/pull/998/commits/43d3d3a2a4df5c76f8255006d8873f7cc63d9953.

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm
/approve
/ok-to-test

**Pull request merged**